rfcs/0042-config-option.md

```diff
@@ -0,0 +1,350 @@
+
+### Additions to the NixOS documentation
+
+The following sections should be added to the NixOS documentation (not verbatim however).
+
+#### Writing options for program configuration
+
+Whether having a structural `settings` option for a module makes sense depends on whether the program's configuration format has an obvious mapping from Nix. This includes formats like JSON, YAML, INI and similar ones. Examples of unsuitable configuration formats are Haskell, Lisp, Lua or other generic programming languages. If you need to ask yourself "Does it make sense to use Nix for this configuration format", then the answer is probably No, and you should not use this approach. This RFC does not specify anything for unsuitable configuration formats, but there is [an addendum on that][unsuitable] regarding this.
+
+The two main ingredients for writing a `settings` option is to define its type as the one corresponding to the programs configuration format (e.g. `lib.formats.json.type`), and to convert that options value to a string with the corresponding generator function (e.g. `lib.formats.json.generate`). Since these options won't include documentation for all supported values, upstream docs should be linked in the description.
+
+#### Default values
+
+Ideally modules should work by just setting `enable = true`, which often requires setting some default settings. These should get specified in the `config` section of the module by assigning the values to the `settings` option directly. Depending on how default settings matter, we need to set them differently and for different reasons:
+
+| Reason | How to assign | Needs to track upstream | Examples | Note |
+| --- | --- | --- | --- | --- |
+| Program would fail otherwise | `mkDefault` | No | `bootstrap_ip = "172.22.68.74"` | Equivalent to a starter configuration |
+| Needed for the module to work, NixOS specifics | **Without** `mkDefault` | No | `logger = "systemd"` `data_dir = "/var/lib/foo"` | Requires the user to use `mkForce` for overriding this, hinting that they leave supported territory |
+| Module needs value to influence other options | `mkDefault` | Yes | `port = 456` (influences `allowedTCPPorts`) | |
+
+#### Additional options for single settings
+
+One can easily add additional options that correspond to single configuration settings. This is done by defining an option as usual, then applying it to `settings` with a `mkDefault`. This approach allows users to set the value either through the specialized option or `settings`, which also means that new options can be added without having to worry about backwards-compatibility.
+
+#### An example
+
+Putting it all together, here is an example of a NixOS module that uses such an approach:
+
+```nix
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.services.foo;
+  format = lib.formats.json;
```

Comment:
```suggestion
  format = pkgs.formats.json;
```

Reply:
Although I feel this scope is a bit unfortunate, barring any future `formats` package from entering top-level.

Reply:
Reference: https://github.com/NixOS/nixpkgs/pull/75584/files#diff-036410e9211b4336186fc613f7200b12